securedrop/crypto_util.py

```diff
@@ -1,54 +1,207 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
 
-from base64 import b32encode
+import gnupg
 import os
+import scrypt
 import subprocess
 
+from base64 import b32encode
 from Cryptodome.Random import random
-import gnupg
+from flask import current_app
 from gnupg._util import _is_stream, _make_binary_stream
-import scrypt
 from typing import Dict, List, Text  # noqa: F401
 
-import config
-import store
-
 # to fix gpg error #78 on production
 os.environ['USERNAME'] = 'www-data'
 
-GPG_KEY_TYPE = "RSA"
-if os.environ.get('SECUREDROP_ENV') == 'test':
-    # Optimize crypto to speed up tests (at the expense of security - DO NOT
-    # use these settings in production)
-    GPG_KEY_LENGTH = 1024
-    SCRYPT_PARAMS = dict(N=2**1, r=1, p=1)
-else:  # pragma: no cover
-    GPG_KEY_LENGTH = 4096
-    SCRYPT_PARAMS = config.SCRYPT_PARAMS
-
-SCRYPT_ID_PEPPER = config.SCRYPT_ID_PEPPER
-SCRYPT_GPG_PEPPER = config.SCRYPT_GPG_PEPPER
-
-DEFAULT_WORDS_IN_RANDOM_ID = 8
-
-
-# Make sure these pass before the app can run
-# TODO: Add more tests
-def do_runtime_tests():
-    if config.SCRYPT_ID_PEPPER == config.SCRYPT_GPG_PEPPER:
-        raise AssertionError('SCRYPT_ID_PEPPER == SCRYPT_GPG_PEPPER')
-    # crash if we don't have srm:
-    try:
-        subprocess.check_call(['srm'], stdout=subprocess.PIPE)
-    except subprocess.CalledProcessError:
-        pass
 
+class CryptoException(Exception):
+    pass
 
-do_runtime_tests()
 
-gpg = gnupg.GPG(binary='gpg2', homedir=config.GPG_KEY_DIR)
+class CryptoUtil:
+
+    GPG_KEY_TYPE = "RSA"
+    DEFAULT_WORDS_IN_RANDOM_ID = 8
+
+    def __init__(self,
+                 scrypt_params,
+                 scrypt_id_pepper,
+                 scrypt_gpg_pepper,
+                 securedrop_root,
+                 word_list,
+                 nouns_file,
+                 adjectives_file,
+                 gpg_key_dir):
+        self.__securedrop_root = securedrop_root
+        self.__word_list = word_list
+
+        if os.environ.get('SECUREDROP_ENV') == 'test':
+            # Optimize crypto to speed up tests (at the expense of security
+            # DO NOT use these settings in production)
+            self.__gpg_key_length = 1024
+            self.scrypt_params = dict(N=2**1, r=1, p=1)
+        else:  # pragma: no cover
+            self.__gpg_key_length = 4096
+            self.scrypt_params = scrypt_params
+
+        self.scrypt_id_pepper = scrypt_id_pepper
+        self.scrypt_gpg_pepper = scrypt_gpg_pepper
+
+        self.do_runtime_tests()
+
+        self.__gpg_key_dir = gpg_key_dir
+        self.gpg = gnupg.GPG(binary='gpg2', homedir=gpg_key_dir)
+
+        # map code for a given language to a localized wordlist
+        self.__language2words = {}  # type: Dict[Text, List[str]]
+
+        with open(nouns_file) as f:
```

Comment:
Thank you for these context managers.  💯

Reply:
We should try to move all the `open` calls into `io.open` calls. This will help us to another step towards Python3.

Reply:
But why? Python3 has `open` as a function?

Reply:
Because then we will be using the same function call in Python2 and in future in Python3. The `open` call in Python3 actually points to `io.open` 

If we keep using `open` function call now, means we will change the implementation when we switch to Python3. The suggestion we provide is to use `io.open` when a project tries to move towards Python3.

Reply:
Ok, well we can do that in another PR. This is a restructuring of the current code to allow us to do dependency injection.

Reply:
Okay.